toolchain/diagnostics/check_diagnostics.py

```diff
@@ -0,0 +1,116 @@
+#!/usr/bin/env python3
+
+"""Checks diagnostic use.
+
+Validates that each diagnostic declared with CARBON_DIAGNOSTIC_KIND is
+referenced by one (and only one) CARBON_DIAGNOSTIC.
+"""
+
+__copyright__ = """
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+"""
+
+from concurrent import futures
+import itertools
+from pathlib import Path
+import os
+import re
+import sys
+from typing import Dict, List, NamedTuple, Set
+
+
+# Example or test diagnostics, ignored because they're expected to not pass.
+IGNORED = set(["MyDiagnostic", "TestDiagnostic"])
+
+
+class Location(NamedTuple):
+    """A location for a diagnostic."""
+
+    def __str__(self) -> str:
+        return f"{str(self.path)}:{self.line}"
+
+    path: Path
+    line: int
+
+
+def load_diagnostic_kind() -> Set[str]:
+    """Returns the set of declared diagnostic kinds.
+
+    This isn't validated for uniqueness because the compiler does that.
+    """
+    path = Path("toolchain/diagnostics/diagnostic_registry.def")
+    content = path.read_text()
+    decls = set(re.findall(r"CARBON_DIAGNOSTIC_KIND\((.+)\)", content))
+    return decls.difference(IGNORED)
+
+
+def load_diagnostic_uses_in(
+    path: Path,
+) -> Dict[str, List[Location]]:
+    """Returns the path's CARBON_DIAGNOSTIC uses."""
+    content = path.read_text()
+
+    # Keep a line cursor so that we don't keep re-scanning the file.
+    line = 1
+    line_offset = 0
+
+    found: Dict[str, List[Location]] = {}
+    for m in re.finditer(r"CARBON_DIAGNOSTIC\(\s*(\w+),", content):
+        diag = m.group(1)
+        if diag in IGNORED:
+            continue
+        line += content.count("\n", line_offset, m.start())
+        line_offset = m.start()
+        found.setdefault(diag, []).append(Location(path, line))
```

Comment:
Would a `defaultdict` make more sense instead of a `Dict`+`setdefault`?

Reply:
Sure, done